core/src/main/java/org/ehcache/core/Ehcache.java

```diff
@@ -1031,4 +969,116 @@ public V getValue() {
+    private final AtomicInteger actualPutCount = new AtomicInteger();
+
+    public PutAllFunction(Logger logger, Map<K, V> entriesToRemap, Expiry<? super K, ? super V> expiry) {
+      this.logger = logger;
+      this.entriesToRemap = entriesToRemap;
+      this.expiry = expiry;
+    }
+
+    @Override
+    public Iterable<? extends Map.Entry<? extends K, ? extends V>> apply(final Iterable<? extends Map.Entry<? extends K, ? extends V>> entries) {
+      Map<K, V> mutations = new LinkedHashMap<K, V>();
+
+      // then record we handled these mappings
+      for (Map.Entry<? extends K, ? extends V> entry: entries) {
+        K key = entry.getKey();
+        V existingValue = entry.getValue();
+        V newValue = entriesToRemap.remove(key);
+
+        if (newValueAlreadyExpired(key, existingValue, newValue)) {
+          mutations.put(key, null);
+        } else {
+          actualPutCount.incrementAndGet();
+          mutations.put(key, newValue);
+        }
+      }
+
+      // Finally return the values to be installed in the Cache's Store
+      return mutations.entrySet();
+    }
+
+    public Map<K, V> getEntriesToRemap() {
+      return entriesToRemap;
+    }
+
+    private boolean newValueAlreadyExpired(K key, V oldValue, V newValue) {
+      if (newValue == null) {
+        return false;
+      }
+
+      final Duration duration;
+      if (oldValue == null) {
+        try {
+          duration = expiry.getExpiryForCreation(key, newValue);
+        } catch (RuntimeException re) {
+          logger.error("Expiry computation caused an exception - Expiry duration will be 0 ", re);
+          return true;
+        }
+      } else {
+        try {
+          duration = expiry.getExpiryForUpdate(key, supplierOf(oldValue), newValue);
+        } catch (RuntimeException re) {
+          logger.error("Expiry computation caused an exception - Expiry duration will be 0 ", re);
+          return true;
+        }
+      }
+
+      return Duration.ZERO.equals(duration);
+    }
+
+    public AtomicInteger getActualPutCount() {
+      return actualPutCount;
+    }
+  }
+
+  public static class RemoveAllFunction<K, V> implements Function<Iterable<? extends Map.Entry<? extends K, ? extends V>>, Iterable<? extends Map.Entry<? extends K, ? extends V>>> {
+
+    private final AtomicInteger actualRemoveCount = new AtomicInteger();
+
+    @Override
+    public Iterable<? extends Map.Entry<? extends K, ? extends V>> apply(final Iterable<? extends Map.Entry<? extends K, ? extends V>> entries) {
+      Map<K, V> results = new LinkedHashMap<K, V>();
+
+      for (Map.Entry<? extends K, ? extends V> entry : entries) {
+        K key = entry.getKey();
+        V existingValue = entry.getValue();
+
+        if (existingValue != null) {
+          actualRemoveCount.incrementAndGet();
+        }
+        results.put(key, null);
+      }
+
+      return results.entrySet();
+    }
+
+    public AtomicInteger getActualRemoveCount() {
```

Comment:
Same as above: I'd make this method return `int` even if you keep the `AtomicInteger`.